File: cmajor/parsing/Nonterminal.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Nonterminal.hpp>
#include <cmajor/parsing/Action.hpp>
#include <cmajor/parsing/Rule.hpp>
#include <cmajor/parsing/Visitor.hpp>

namespace cmajor { namespace parsing {

NonterminalParser::NonterminalParser(const std::u32string& name_, const std::u32string& ruleName_):
    Parser(name_, U"<" + name_ + U">"), name(name_), ruleName(ruleName_), numberOfArguments(0), preCall(), postCall()
{
}

NonterminalParser::NonterminalParser(const std::u32string& name_, const std::u32string& ruleName_, int numberOfArguments_):
    Parser(name_, U"<" + name_ + U">"), name(name_), ruleName(ruleName_), numberOfArguments(numberOfArguments_), preCall(), postCall()
{
}

void NonterminalParser::SetPreCall(PreCall* preCall_) 
{ 
    preCall = std::unique_ptr<PreCall>(preCall_); 
}

void NonterminalParser::SetPostCall(PostCall* postCall_) 
{ 
    postCall = std::unique_ptr<PostCall>(postCall_); 
}

void NonterminalParser::SetArguments(const ArgumentVector& arguments_)
{
    arguments = arguments_; 
    for (cmajor::codedom::CppObject* arg : arguments)
    {
        if (!arg->IsOwned())
        {
            arg->SetOwned();
            ownedArgs.push_back(std::unique_ptr<cmajor::codedom::CppObject>(arg));
        }
    }
}

bool NonterminalParser::Specialized() const 
{ 
    return numberOfArguments > 0 || !arguments.empty() || 
        (rule && rule->TakesOrReturnsValuesOrHasLocals()); 
}

Match NonterminalParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    if (rule)
    {
        if (preCall)
        {
            (*preCall)(stack, parsingData);
        }
        Match match = rule->Parse(scanner, stack, parsingData);
        if (postCall)
        {
            (*postCall)(stack, parsingData, match.Hit());
        }
        return match;
    }
    return Match::Nothing();
}

void NonterminalParser::Accept(Visitor& visitor)
{
    visitor.Visit(*this);
}

} } // namespace cmajor::parsing

File: devenv/soul.parsing/NonTerminal.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using soul.codedom;

namespace soul.parsing
{
    public delegate void PreCall(Stack<object> stack);
    public delegate void PostCall(Stack<object> stack, bool matched);

    public class NonTerminalParser : Parser, INamedEntity
    {
        public NonTerminalParser(string name, string fullName, string ruleName, int numberOfArguments)
            : base("<" + fullName + ">")
        {
            this.name = name;
            this.fullName = fullName;
            this.ruleName = ruleName;
            this.preCall = null;
            this.postCall = null;
            this.numberOfArguments = numberOfArguments;
        }
        public NonTerminalParser(string name, string fullName, string ruleName, List<CSharpEntity> arguments)
            : base("<" + fullName + ">")
        {
            this.name = name;
            this.fullName = fullName;
            this.ruleName = ruleName;
            this.arguments = arguments;
            this.preCall = null;
            this.postCall = null;
            this.numberOfArguments = arguments != null ? arguments.Count : 0;
        }
        public string Name
        {
            get { return name; }
        }
        public string FullName
        {
            get { return fullName; }
        }
        public string RuleName
        {
            get { return ruleName; }
        }
        public bool Specialized
        {
            get { return numberOfArguments > 0 || name != fullName || rule.TakesOrReturnsValuesOrHasLocals; }
        }
        public int NumberOfArguments
        {
            get { return numberOfArguments; }
        }
        public List<CSharpEntity> Arguments
        {
            get { return arguments; }
            set { arguments = value; }
        }
        public Position Position
        {
            get { return position; }
            set { position = value; }
        }
        public RuleParser Rule
        {
            get { return rule; }
            set { rule = value; }
        }
        public PreCall PreCall
        {
            get { return preCall; }
            set { preCall = value; }
        }
        public PostCall PostCall
        {
            get { return postCall; }
            set { postCall = value; }
        }
        public string PreCallMethodName
        {
            get { return "Pre" + name; }
        }
        public string PostCallMethodName
        {
            get { return "Post" + name; }
        }
        public string ValueFieldName
        {
            get { return "from" + name; }
        }
        public string VariableName
        {
            get { return (name.Length > 0 ? char.ToLower(name[0]) + name.Substring(1) : name) + "NonTerminalParser"; }
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            if (rule != null)
            {
                if (preCall != null)
                {
                    preCall(stack);
                }
                Match match = rule.Parse(scanner, stack);
                if (postCall != null)
                {
                    postCall(stack, match.Hit);
                }
                return match;
            }
            return Match.Nothing();
        }
        public override void Accept(Visitor visitor)
        {
            visitor.Visit(this);
        }
        private string name;
        private string fullName;
        private string ruleName;
        private RuleParser rule;
        private int numberOfArguments;
        private List<CSharpEntity> arguments;
        private PreCall preCall;
        private PostCall postCall;
        private Position position;
    }
}
